### PR TITLE
[Feature] Support spark job status tracking

### PR DESCRIPTION
**Pull request by lenoxzhao**
## Brief change log
1. Modify the spark job submission flow on yarn. Yarn application id will be carried on `SubmitResponse` for later task tracking.
2. Support spark parameters configuring.
image
3. Add `SparkAppHttpWatcher` refer to this FlinkAppHttpWatcher. Application state can be accessed from yarn rest api,  resource can be accessed from spark rest api (only available when job is in RUNNING state).
image2
**TODO: Now the information being tracked will be temporarily treated as log, and will be persisted in the future when the new spark database table is designed (now sharing the table with flink).**
4. Simplify spark application state enumeration according to [YarnApplicationState](https://hadoop.apache.org/docs/stable/hadoop-yarn/hadoop-yarn-api/apidocs/org/apache/hadoop/yarn/api/records/YarnApplicationState.html).
5. Change "cancel" operation to "stop" operation.
6. Removed some flink-only processes.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

## Does this pull request potentially affect one of the following parts
 - Dependencies (does it add or upgrade a dependency): (no)

### CODE REVIEW

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/base/config/AsyncExecutorPoolConfig.java [hunk 1]
```diff
@@ -66,6 +66,22 @@ public Executor restAPIWatchingExecutor() {
                 ThreadUtils.threadFactory("flink-restapi-watching-executor-"));
     }
 
+    /**
+     * Create a ThreadPoolTaskExecutor for FlinkClusterWatcher.
```
Comment (ChengJie1053) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/base/config/AsyncExecutorPoolConfig.java]: flink should be changed to spark

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+    /** tracking task list */
+    private static final Map<Long, SparkApplication> WATCHING_APPS = new ConcurrentHashMap<>(0);
+
+    /**
+     *
+     *
+     * <pre>
+     * StopFrom: marked a flink job canceling from the StreamPark or other ways:
+     *    1) If stop from streampark, We can know whether to make a savepoint when flink job
+     * canceling, and if We make a savepoint,
```
Comment (ChengJie1053) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java]: flink should be changed to spark

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+     *
+     * <p>If the Flink cluster configuration lacks alarm information, it triggers an alarm directly
+     * when the job is abnormal.<br>
+     * If the Flink cluster configuration has alarm information:
+     *
+     * <p>When the job is abnormal due to an issue in the Flink cluster, the job's alarm will be held
+     * back, instead waiting for the Flink cluster's alarm.<br>
+     * When the job is abnormal due to the job itself and the Flink cluster is running normally, an
+     * alarm specific to the job will be triggered.
+     *
```
Comment (ChengJie1053) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java:381]: flink should be changed to spark

File: streampark-spark/streampark-spark-client/streampark-spark-client-core/src/main/scala/org/apache/streampark/spark/client/trait/SparkClientTrait.scala [hunk 4]
```diff
@@ -38,7 +38,7 @@ trait SparkClientTrait extends Logger {
          |    properties       : ${submitRequest.properties.mkString(" ")}
          |    args             : ${submitRequest.args}
          |    appConf          : ${submitRequest.appConf}
-         |-------------------------------------------------------------------------------------------
+         |------------------------------------------------------------------------------`-------------
```
Comment (ChengJie1053) [inline streampark-spark/streampark-spark-client/streampark-spark-client-core/src/main/scala/org/apache/streampark/spark/client/trait/SparkClientTrait.scala]: This line should not be changed

File: streampark-spark/streampark-spark-client/streampark-spark-client-api/src/main/scala/org/apache/streampark/spark/client/conf/SparkConfiguration.scala [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.streampark.spark.client.conf
+
+object SparkConfiguration {
+  val defaultParameters = Map[String, Any](
+    "spark.driver.cores" -> "1",
```
Comment (wolfboys) [inline streampark-spark/streampark-spark-client/streampark-spark-client-api/src/main/scala/org/apache/streampark/spark/client/conf/SparkConfiguration.scala:22]: Can these parameters be customized by users?

Comment (lenoxzhao) [reply streampark-spark/streampark-spark-client/streampark-spark-client-api/src/main/scala/org/apache/streampark/spark/client/conf/SparkConfiguration.scala:22]: Yes. Users can customize parameters through `DynamicPropeties`. The user-defined parameters will override the default parameters before the job is submitted.

```java
  //streampark-spark/streampark-spark-client/streampark-spark-client-core/src/main/scala/org/apache/streampark/spark/client/impl/YarnApplicationClient.scala#setDynamicProperties
  private def setDynamicProperties(sparkLauncher: SparkLauncher, properties: Map[String, Any]): Unit = {
    logger.info("Spark launcher start configuration.")
    val finalProperties: Map[String, Any] = SparkConfiguration.defaultParameters ++ properties
    for ((k, v) <- finalProperties) {
      if (k.startsWith("spark.")) {
        sparkLauncher.setConf(k, v.toString)
      } else {
        logger.info("\"{}\" doesn't start with \"spark.\". Skip it.", k)
      }
    }
  }
```

Comment (wolfboys) [reply streampark-spark/streampark-spark-client/streampark-spark-client-api/src/main/scala/org/apache/streampark/spark/client/conf/SparkConfiguration.scala:22]: good job

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/enums/SparkOptionStateEnum.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+@Getter
+public enum SparkOptionStateEnum {
+
+    /** Application which is currently action: none. */
+    NONE(0),
+    /** Application which is currently action: releasing. */
+    RELEASING(1),
+    /** Application which is currently action: starting. */
+    STARTING(2),
+    /** Application which is currently action: starting. */
```
Comment (ChengJie1053) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/enums/SparkOptionStateEnum.java]: /** Application which is currently action:stopping. */

File: streampark-spark/streampark-spark-client/streampark-spark-client-core/src/main/scala/org/apache/streampark/spark/client/impl/YarnApplicationClient.scala [hunk 7]
```diff
@@ -58,67 +58,52 @@ object YarnApplicationClient extends SparkClientTrait {
       .setMaster("yarn")
       .setDeployMode("cluster")
```
Comment (ChengJie1053) [inline streampark-spark/streampark-spark-client/streampark-spark-client-core/src/main/scala/org/apache/streampark/spark/client/impl/YarnApplicationClient.scala]: DeployMode can set to 'cluster' and 'client'

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/SparkApplicationLog.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+@Slf4j
+public class SparkApplicationLog {
+
+    @TableId(type = IdType.AUTO)
+    private Long id;
+    /** appId */
+    private Long appId;
+    /** applicationId */
+    private String sparkAppId;
+    /** The address of the jobmanager, that is, the direct access address of the Flink web UI */
```
Comment (ChengJie1053) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/SparkApplicationLog.java]: modify code comment

File: streampark-spark/streampark-spark-client/streampark-spark-client-api/src/main/scala/org/apache/streampark/spark/client/proxy/SparkShimsProxy.scala [hunk 9]
```diff
@@ -140,7 +140,7 @@ object SparkShimsProxy extends Logger {
     SHIMS_CLASS_LOADER_CACHE.getOrElseUpdate(
       s"${sparkVersion.fullVersion}", {
         // 1) spark/lib
-        val libURL = getSparkHomeLib(sparkVersion.sparkHome, "jars", !_.getName.startsWith("log4j"))
+        val libURL = getSparkHomeLib(sparkVersion.sparkHome, "jars", f => !f.getName.startsWith("log4j") && !f.getName.startsWith("slf4j"))
```
Comment (ChengJie1053) [inline streampark-spark/streampark-spark-client/streampark-spark-client-api/src/main/scala/org/apache/streampark/spark/client/proxy/SparkShimsProxy.scala]: `libURL` changed to `libUrl`

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+     */
+    private static final Cache<Long, Byte> STARTING_CACHE =
+        Caffeine.newBuilder().expireAfterWrite(5, TimeUnit.MINUTES).build();
+
+    /** tracking task list */
+    private static final Map<Long, SparkApplication> WATCHING_APPS = new ConcurrentHashMap<>(0);
+
+    /**
+     *
+     *
```
Comment (ChengJie1053) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java]: Remove redundant comments

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java [hunk 11]
```diff
@@ -0,0 +1,415 @@
...
+     * StopFrom: Recording spark application stopped by streampark or stopped by other actions
+     * </pre>
+     */
+    private static final Map<Long, StopFromEnum> STOP_FROM_MAP = new ConcurrentHashMap<>(0);
+
+    /**
+     * Cancelling tasks are placed in this cache with an expiration time of 10 seconds (the time of 2
+     * task monitoring polls).
+     */
+    private static final Cache<Long, Byte> CANCELING_CACHE =
```
Comment (ChengJie1053) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java]: CANCELING_CACHE  is not used

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java [hunk 12]
```diff
@@ -0,0 +1,415 @@
...
+    private <T> T yarnRestRequest(String url, Class<T> clazz) throws IOException {
+        String result = YarnUtils.restRequest(url, HTTP_TIMEOUT);
+        if (null == result) {
+            return null;
+        }
+        return JacksonUtils.read(result, clazz);
+    }
+
+    private <T> T httpRestRequest(String url, Class<T> clazz) throws IOException {
+        String result =
```
Comment (ChengJie1053) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java]: httpRestRequest  is not used

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/SparkAppBuildPipeServiceImpl.java [hunk 13]
```diff
@@ -129,7 +129,7 @@ public class SparkAppBuildPipeServiceImpl
     private SparkApplicationInfoService applicationInfoService;
 
     @Autowired
-    private ApplicationLogService applicationLogService;
+    private SparkApplicationLogService applicationLogService;
 
     @Autowired
     private FlinkAppHttpWatcher flinkAppHttpWatcher;
```
Comment (ChengJie1053) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/SparkAppBuildPipeServiceImpl.java]: Whether you should use SparkAppHttpWatcher

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/application/impl/SparkApplicationInfoServiceImpl.java [hunk 14]
```diff
@@ -35,15 +35,16 @@
...
 import org.apache.streampark.console.core.enums.FlinkAppStateEnum;
+import org.apache.streampark.console.core.enums.SparkAppStateEnum;
 import org.apache.streampark.console.core.mapper.SparkApplicationMapper;
 import org.apache.streampark.console.core.metrics.flink.JobsOverview;
 import org.apache.streampark.console.core.runner.EnvInitializer;
 import org.apache.streampark.console.core.service.FlinkClusterService;
-import org.apache.streampark.console.core.service.SavePointService;
 import org.apache.streampark.console.core.service.SparkEnvService;
 import org.apache.streampark.console.core.service.application.SparkApplicationInfoService;
 import org.apache.streampark.console.core.watcher.FlinkAppHttpWatcher;
```
Comment (ChengJie1053) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/application/impl/SparkApplicationInfoServiceImpl.java]: Whether you should use SparkAppHttpWatcher

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java [hunk 15]
```diff
@@ -0,0 +1,415 @@
...
+        }
+        return JacksonUtils.read(result, clazz);
+    }
+
+    public boolean isWatchingApp(Long id) {
+        return WATCHING_APPS.containsKey(id);
+    }
+
+    interface Callback<T, R> {
+
```
Comment (ChengJie1053) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java]: Callback is not used

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/base/config/AsyncExecutorPoolConfig.java (hunk 1 above)
Comment (lenoxzhao) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/base/config/AsyncExecutorPoolConfig.java]: done

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java (hunk 2 above)
Comment (lenoxzhao) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java]: done

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java (hunk 3 above)
Comment (lenoxzhao) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java:381]: done

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java (hunk 15 above)
Comment (lenoxzhao) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java]: done

File: streampark-spark/streampark-spark-client/streampark-spark-client-core/src/main/scala/org/apache/streampark/spark/client/impl/YarnApplicationClient.scala (hunk 7 above)
Comment (lenoxzhao) [reply streampark-spark/streampark-spark-client/streampark-spark-client-core/src/main/scala/org/apache/streampark/spark/client/impl/YarnApplicationClient.scala]: done

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/enums/SparkOptionStateEnum.java (hunk 6 above)
Comment (lenoxzhao) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/enums/SparkOptionStateEnum.java]: done

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/SparkAppBuildPipeServiceImpl.java (hunk 13 above)
Comment (lenoxzhao) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/SparkAppBuildPipeServiceImpl.java]: done

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java (hunk 12 above)
Comment (lenoxzhao) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java]: Already removed.

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/SparkApplicationLog.java (hunk 8 above)
Comment (lenoxzhao) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/entity/SparkApplicationLog.java]: done

File: streampark-spark/streampark-spark-client/streampark-spark-client-api/src/main/scala/org/apache/streampark/spark/client/proxy/SparkShimsProxy.scala (hunk 9 above)
Comment (lenoxzhao) [reply streampark-spark/streampark-spark-client/streampark-spark-client-api/src/main/scala/org/apache/streampark/spark/client/proxy/SparkShimsProxy.scala]: done

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java (hunk 10 above)
Comment (lenoxzhao) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java]: done

File: streampark-spark/streampark-spark-client/streampark-spark-client-core/src/main/scala/org/apache/streampark/spark/client/trait/SparkClientTrait.scala (hunk 4 above)
Comment (lenoxzhao) [reply streampark-spark/streampark-spark-client/streampark-spark-client-core/src/main/scala/org/apache/streampark/spark/client/trait/SparkClientTrait.scala]: done

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java (hunk 11 above)
Comment (lenoxzhao) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java]: Already removed.

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/application/impl/SparkApplicationInfoServiceImpl.java (hunk 14 above)
Comment (lenoxzhao) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/application/impl/SparkApplicationInfoServiceImpl.java]: done

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/application/impl/SparkApplicationActionServiceImpl.java [hunk 16]
```diff
@@ -417,16 +352,17 @@ public void start(SparkApplication appParam, boolean auto) throws Exception {
...
                         updateById(app);
-                        FlinkAppHttpWatcher.unWatching(appParam.getId());
+                        SparkAppHttpWatcher.unWatching(appParam.getId());
                     }
                     return;
                 }
 
                 // 3) success
                 applicationLog.setSuccess(true);
+                // TODO：修改为spark对应的参数
```
Comment (ChengJie1053) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/application/impl/SparkApplicationActionServiceImpl.java:365]: Does this place need to be modified or modified in another pr

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/application/impl/SparkApplicationActionServiceImpl.java [hunk 17]
```diff
@@ -508,8 +444,7 @@ private Tuple2<String, String> getUserJarAndAppConf(
                 FlinkSql flinkSql = flinkSqlService.getEffective(application.getId(), false);
                 AssertUtils.notNull(flinkSql);
```
Comment (ChengJie1053) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/application/impl/SparkApplicationActionServiceImpl.java:445]: Does sparksql need to be distinguished from FlinkSql and stored in a different table

Comment (ChengJie1053) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/application/impl/SparkApplicationActionServiceImpl.java:445]: image
Does it need to be modified here

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java [hunk 18]
```diff
@@ -0,0 +1,389 @@
...
+                    double maxMemory = resourceStatus.t2;
+                    double totalCores = resourceStatus.t3;
+                    log.info(
+                        "[StreamPark][SparkAppHttpWatcher] getStateFromYarn, app {} was running, jobId is {}, memoryUsed: {}MB, maxMemory: {}MB, totalCores: {}",
+                        application.getId(),
+                        application.getJobId(),
+                        String.format("%.2f", memoryUsed),
+                        String.format("%.2f", maxMemory),
+                        totalCores);
+                    // TODO: Modify the table structure to persist the results
```
Comment (ChengJie1053) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java:226]: Do you need to delete?

File: streampark-spark/streampark-spark-client/streampark-spark-client-core/src/main/scala/org/apache/streampark/spark/client/impl/YarnApplicationClient.scala [hunk 19]
```diff
@@ -18,107 +18,100 @@
...
-          .asInstanceOf[SubmitRequest]
           .hdfsWorkspace
           .sparkLib + "/*.jar")
       .setVerbose(true)
 
+    import scala.collection.JavaConverters._
+    setDynamicProperties(launcher, submitRequest.properties.asScala.toMap)
+
+    // TODO: Adds command line arguments for the application.
+    // launcher.addAppArgs()
```
Comment (ChengJie1053) [inline streampark-spark/streampark-spark-client/streampark-spark-client-core/src/main/scala/org/apache/streampark/spark/client/impl/YarnApplicationClient.scala:80]: Whether this place needs to be modified？

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java [hunk 20]
```diff
@@ -0,0 +1,389 @@
...
+                try {
+                    getStateFromYarn(application);
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+            });
+    }
+
+    private StopFromEnum getAppStopFrom(Long appId) {
+        return STOP_FROM_MAP.getOrDefault(appId, StopFromEnum.NONE);
```
Comment (ChengJie1053) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java:184]: `getAppStopFrom` is not used

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java [hunk 21]
```diff
@@ -0,0 +1,389 @@
...
+    /**
+     * Calculate spark task progress from Spark rest api. (proxyed by yarn) Only available when yarn
+     * application status is RUNNING.
+     *
+     * @param application
+     * @return task progress
+     * @throws Exception
+     */
+    private double getTasksProgress(SparkApplication application) throws Exception {
+        Job[] jobs = httpJobsStatus(application);
```
Comment (ChengJie1053) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/watcher/SparkAppHttpWatcher.java:254]: `getTasksProgress`  is not used

**Review by wolfboys — APPROVED** [review]
LGTM
